### PR TITLE
kucoinfutures: add stopLossPrice and takeProfitPrice support

### PR DESCRIPTION
**Pull request by Dan-krm**
Added stopLossPrice and takeProfitPrice support to kucoinfutures: order type 3 is not currently supported by the kucoin futures API https://github.com/ccxt/ccxt/issues/13822
### Stop Trigger Order:
```
node examples/js/cli kucoinfutures createOrder BTC/USDT:USDT limit buy 1 13000 '{"triggerPrice":"14000"}'

{
  id: '63b36a10db691700019e5d14',
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  symbol: undefined,
  type: undefined,
  side: undefined,
  price: undefined,
  amount: undefined,
  cost: undefined,
  average: undefined,
  filled: undefined,
  remaining: undefined,
  status: undefined,
  fee: undefined,
  trades: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  info: { code: '200000', data: { orderId: '63b36a10db691700019e5d14' } }
}
```

### Stoploss:
```
node examples/js/cli kucoinfutures createOrder BTC/USDT:USDT limit sell 1 14500 '{"stopLossPrice":"15000"}'

{
  id: '63b36ba6db691700019f8be9',
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  symbol: undefined,
  type: undefined,
  side: undefined,
  price: undefined,
  amount: undefined,
  cost: undefined,
  average: undefined,
  filled: undefined,
  remaining: undefined,
  status: undefined,
  fee: undefined,
  trades: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  info: { code: '200000', data: { orderId: '63b36ba6db691700019f8be9' } }
}
```

### Takeprofit:
```
node examples/js/cli kucoinfutures createOrder BTC/USDT:USDT limit sell 1 17500 '{"takeProfitPrice":"17000"}'

{
  id: '63b36bf6315843000178c894',
  clientOrderId: undefined,
  timestamp: undefined,
  datetime: undefined,
  lastTradeTimestamp: undefined,
  symbol: undefined,
  type: undefined,
  side: undefined,
  price: undefined,
  amount: undefined,
  cost: undefined,
  average: undefined,
  filled: undefined,
  remaining: undefined,
  status: undefined,
  fee: undefined,
  trades: undefined,
  timeInForce: undefined,
  postOnly: undefined,
  stopPrice: undefined,
  triggerPrice: undefined,
  info: { code: '200000', data: { orderId: '63b36bf6315843000178c894' } }
}
```

### CODE REVIEW

File: js/kucoinfutures.js [hunk 1]
```diff
@@ -1059,11 +1064,38 @@ module.exports = class kucoinfutures extends kucoin {
...
+        const takeProfitPrice = this.safeValue (params, 'takeProfitPrice');
+        const isStopLoss = stopLossPrice !== undefined;
+        const isTakeProfit = takeProfitPrice !== undefined;
+        const stop = this.safeString (params, 'stop');
+        const stopPriceType = this.safeString (params, 'stopPriceType', 'MP');
         if (stopPrice) {
-            request['stop'] = (side === 'buy') ? 'up' : 'down';
-            const stopPriceType = this.safeString (params, 'stopPriceType', 'TP');
-            request['stopPriceType'] = stopPriceType;
+            if (stop !== undefined) {
```
Comment (carlosmiei) [inline js/kucoinfutures.js:1074]: again we don't need to do this `request` will be extended with `params`, so you can replace this `if {} else{}` with 
```Javascript
if (stop === undefined) { 
      request['stop'] = (side === 'buy') ? 'down' : 'up';
}
```

**Comment by carlosmiei** [comment]
LGTM ✅ 
```
 p kucoinfutures createOrder "BTC/USDT:USDT" "limit" "buy" 1 14000 '{"stopLossPrice":10000}' --sandbox
Python v3.10.8
CCXT v2.5.12
kucoinfutures.createOrder(BTC/USDT:USDT,limit,buy,1,14000,{'stopLossPrice': 10000})
{'amount': None,
 'average': None,
 'clientOrderId': None,
 'cost': None,
 'datetime': None,
 'fee': None,
 'filled': None,
 'id': '63b6c5b62b968a0001b5300b',
 'info': {'code': '200000', 'data': {'orderId': '63b6c5b62b968a0001b5300b'}},
 'lastTradeTimestamp': None,
 'postOnly': None,
 'price': None,
 'remaining': None,
 'side': None,
 'status': None,
 'stopPrice': None,
 'symbol': None,
 'timeInForce': None,
 'timestamp': None,
 'trades': None,
 'triggerPrice': None,
 'type': None}
```
```
 p kucoinfutures createOrder "BTC/USDT:USDT" "limit" "buy" 1 14000 '{"takeProfitPrice":10000}' --sandbox          
Python v3.10.8
CCXT v2.5.32
kucoinfutures.createOrder(BTC/USDT:USDT,limit,buy,1,14000,{'takeProfitPrice': 10000})
{'amount': None,
 'average': None,
 'clientOrderId': None,
 'cost': None,
 'datetime': None,
 'fee': None,
 'filled': None,
 'id': '63b6c81ff1ee3000013e67fb',
 'info': {'code': '200000', 'data': {'orderId': '63b6c81ff1ee3000013e67fb'}},
 'lastTradeTimestamp': None,
 'postOnly': None,
 'price': None,
 'remaining': None,
 'side': None,
 'status': None,
 'stopPrice': None,
 'symbol': None,
 'timeInForce': None,
 'timestamp': None,
 'trades': None,
 'triggerPrice': None,
 'type': None}
```

**Review by carlosmiei — APPROVED** [review]

File: js/kucoinfutures.js [hunk 2]
```diff
@@ -1059,11 +1064,32 @@ module.exports = class kucoinfutures extends kucoin {
...
+        const takeProfitPrice = this.safeValue (params, 'takeProfitPrice');
+        const isStopLoss = stopLossPrice !== undefined;
+        const isTakeProfit = takeProfitPrice !== undefined;
+        const stop = this.safeString (params, 'stop');
+        const stopPriceType = this.safeString (params, 'stopPriceType', 'MP');
         if (stopPrice) {
-            request['stop'] = (side === 'buy') ? 'up' : 'down';
-            const stopPriceType = this.safeString (params, 'stopPriceType', 'TP');
-            request['stopPriceType'] = stopPriceType;
+            if (stop === undefined) {
```
Comment (samgermain) [inline js/kucoinfutures.js:1074]: You actually don't need this, stop will be overwritten if it's passed to params then it's overwritten in `extend`

File: js/kucoinfutures.js (hunk 1 above)
Comment (samgermain) [reply js/kucoinfutures.js]: we actually don't need any check, because it will be overwritten

File: js/kucoinfutures.js [hunk 3]
```diff
@@ -1089,7 +1115,7 @@ module.exports = class kucoinfutures extends kucoin {
                 throw new ArgumentsRequired (this.id + ' createOrder() requires a visibleSize parameter for iceberg orders');
             }
         }
-        params = this.omit (params, [ 'timeInForce', 'stopPrice', 'triggerPrice' ]); // Time in force only valid for limit orders, exchange error when gtc for market orders
+        params = this.omit (params, [ 'timeInForce', 'stop', 'stopPrice', 'stopPriceType', 'triggerPrice', 'stopLossPrice', 'takeProfitPrice' ]); // Time in force only valid for limit orders, exchange error when gtc for market orders
```
Comment (samgermain) [inline js/kucoinfutures.js:1118]: Exchange specific params like `stop` and `stopPriceType` shouldn't be removed, if they are set in params, then the value will be overwritten by `extend`. Take a look at https://github.com/ccxt/ccxt/pull/16619 , it sets a default value of `'MP'` and if the user passes `stopPriceType` to params, then the users value takes priority

File: js/kucoinfutures.js [hunk 4]
```diff
@@ -1059,11 +1064,32 @@ module.exports = class kucoinfutures extends kucoin {
             'leverage': 1,
         };
         const stopPrice = this.safeValue2 (params, 'triggerPrice', 'stopPrice');
+        const stopLossPrice = this.safeValue (params, 'stopLossPrice');
+        const takeProfitPrice = this.safeValue (params, 'takeProfitPrice');
+        const isStopLoss = stopLossPrice !== undefined;
+        const isTakeProfit = takeProfitPrice !== undefined;
+        const stop = this.safeString (params, 'stop');
+        const stopPriceType = this.safeString (params, 'stopPriceType', 'MP');
```
Comment (samgermain) [inline js/kucoinfutures.js:1072]: This also isn't needed, you can just set a default value of `MP` for `stopPriceType`, and then if the user passes `stopPriceType` to params, then the request parameter will be overwritten

File: js/kucoinfutures.js [hunk 5]
```diff
@@ -1059,11 +1064,32 @@ module.exports = class kucoinfutures extends kucoin {
...
         if (stopPrice) {
-            request['stop'] = (side === 'buy') ? 'up' : 'down';
-            const stopPriceType = this.safeString (params, 'stopPriceType', 'TP');
-            request['stopPriceType'] = stopPriceType;
+            if (stop === undefined) {
+                request['stop'] = (side === 'buy') ? 'down' : 'up';
```
Comment (samgermain) [inline js/kucoinfutures.js:1075]: These are backwards, when you set a stop buy order, it's for a short, you're saying that when the price rises above (up) past a certain price, then buy the currency, which will exit the short. I already fixed this in https://github.com/ccxt/ccxt/pull/16619

Comment (Dan-krm) [reply js/kucoinfutures.js:1075]: I think this makes sense for stop orders, I don't think we should change stopLoss and takeProfits though

**Pull request merged**